[2.13alpha28] Card Template Editor Preview Crash: sched.counts - ArrayIndexOutOfBoundsException

* SchedV1, new profile
* Add a new Basic Note
* Bury Card
* Card Browser - Edit Note - Edit Card Template - Preview

Crash:

```
 FATAL EXCEPTION: Thread-6
Process: com.ichi2.anki, PID: 19157
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.ichi2.anki/com.ichi2.anki.CardTemplatePreviewer}: java.lang.ArrayIndexOutOfBoundsException: length=3; index=-2
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2951)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3086)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at androidx.test.espresso.base.Interrogator.a(Interrogator.java:31)
	at androidx.test.espresso.base.UiControllerImpl.a(UiControllerImpl.java:130)
	at androidx.test.espresso.base.UiControllerImpl.a(UiControllerImpl.java:124)
	at androidx.test.espresso.base.UiControllerImpl.a(UiControllerImpl.java:40)
	at androidx.test.espresso.action.MotionEvents.a(MotionEvents.java:68)
	at androidx.test.espresso.action.Tap.a(Tap.java:13)
	at androidx.test.espresso.action.Tap$1.sendTap(Tap.java:2)
	at androidx.test.espresso.action.GeneralClickAction.perform(GeneralClickAction.java:9)
	at androidx.test.espresso.ViewInteraction$SingleExecutionViewAction.perform(ViewInteraction.java:6)
	at androidx.test.espresso.ViewInteraction.a(ViewInteraction.java:140)
	at androidx.test.espresso.ViewInteraction$1.call(ViewInteraction.java:4)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.Handler.handleCallback(Handler.java:873)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:193)
	at android.app.ActivityThread.main(ActivityThread.java:6718)
	at java.lang.reflect.Method.invoke(Method.java)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=3; index=-2
	at com.ichi2.libanki.sched.Sched.counts(Sched.java:142)
	at com.ichi2.anki.AbstractFlashcardViewer.updateScreenCounts(AbstractFlashcardViewer.java:1879)
	at com.ichi2.anki.AbstractFlashcardViewer.updateForNewCard(AbstractFlashcardViewer.java:1863)
	at com.ichi2.anki.AbstractFlashcardViewer.fillFlashcard(AbstractFlashcardViewer.java:2397)
	at com.ichi2.anki.AbstractFlashcardViewer.updateCard(AbstractFlashcardViewer.java:2264)
	at com.ichi2.anki.AbstractFlashcardViewer.displayCardQuestion(AbstractFlashcardViewer.java:2034)
	at com.ichi2.anki.AbstractFlashcardViewer.displayCardQuestion(AbstractFlashcardViewer.java:1996)
	at com.ichi2.anki.CardTemplatePreviewer.displayCardQuestion(CardTemplatePreviewer.java:157)
	at com.ichi2.anki.CardTemplatePreviewer.onCollectionLoaded(CardTemplatePreviewer.java:246)
	at com.ichi2.anki.AnkiActivity.startLoadingCollection(AnkiActivity.java:325)
	at com.ichi2.anki.CardTemplatePreviewer.onCreate(CardTemplatePreviewer.java:91)
	at android.app.Activity.performCreate(Activity.java:7144)
	at android.app.Activity.performCreate(Activity.java:7135)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
	at androidx.test.runner.MonitoringInstrumentation.callActivityOnCreate(MonitoringInstrumentation.java:13)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2931)
	... 25 more
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.java
@@ -195,14 +195,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     private boolean mReplayOnTtsInit = false;
 
     protected static final int MENU_DISABLED = 3;
-    // TODO: Consider extracting to ViewModel
-    // Card counts
-    private SpannableString newCount;
-    private SpannableString lrnCount;
-    private SpannableString revCount;
-
-    // ETA
-    private int eta;
 
     // js api developer contact
     private String mCardSuppliedDeveloperContact  = "";
@@ -230,8 +222,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
      * Variables to hold preferences
      */
     private CardAppearance mCardAppearance;
-    private boolean mPrefHideDueCount;
-    private boolean mPrefShowETA;
     private boolean mPrefShowTopbar;
     private boolean mShowTimer;
     protected boolean mPrefWhiteboard;
@@ -264,7 +254,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
 
     // Preferences from the collection
     private boolean mShowNextReviewTime;
-    private boolean mShowRemainingCardCount;
 
     // Answer card & cloze deletion variables
     private String mTypeCorrect = null;
@@ -288,9 +277,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     private WebView mCardWebView;
     private FrameLayout mCardFrame;
     private FrameLayout mTouchLayer;
-    private TextView mTextBarNew;
-    private TextView mTextBarLearn;
-    private TextView mTextBarReview;
     private TextView mChosenAnswer;
     protected TextView mNext1;
     protected TextView mNext2;
@@ -1507,16 +1493,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
         mPreviewNextCard = findViewById(R.id.preview_next_flashcard);
         mPreviewToggleAnswerText = findViewById(R.id.preview_flip_flashcard);
 
-        mTextBarNew = (TextView) findViewById(R.id.new_number);
-        mTextBarLearn = (TextView) findViewById(R.id.learn_number);
-        mTextBarReview = (TextView) findViewById(R.id.review_number);
-
-        if (!mShowRemainingCardCount) {
-            mTextBarNew.setVisibility(View.GONE);
-            mTextBarLearn.setVisibility(View.GONE);
-            mTextBarReview.setVisibility(View.GONE);
-        }
-
         mCardTimer = (Chronometer) findViewById(R.id.card_time);
 
         mChosenAnswer = (TextView) findViewById(R.id.choosen_answer);
@@ -1589,7 +1565,7 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
         webView.setBackgroundColor(Color.argb(1, 0, 0, 0));
 
         // Javascript interface for calling AnkiDroid functions in webview, see card.js
-        webView.addJavascriptInterface(new JavaScriptFunction(), "AnkiDroidJS");
+        webView.addJavascriptInterface(javaScriptFunction(), "AnkiDroidJS");
         return webView;
     }
 
@@ -1714,14 +1690,9 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     }
 
 
-    private void switchTopBarVisibility(int visible) {
+    protected void switchTopBarVisibility(int visible) {
         if (mShowTimer) {
             mCardTimer.setVisibility(visible);
-        }
-        if (mShowRemainingCardCount) {
-            mTextBarNew.setVisibility(visible);
-            mTextBarLearn.setVisibility(visible);
-            mTextBarReview.setVisibility(visible);
         }
         mChosenAnswer.setVisibility(visible);
     }
@@ -1729,11 +1700,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
 
     protected void initControls() {
         mCardFrame.setVisibility(View.VISIBLE);
-        if (mShowRemainingCardCount) {
-            mTextBarNew.setVisibility(View.VISIBLE);
-            mTextBarLearn.setVisibility(View.VISIBLE);
-            mTextBarReview.setVisibility(View.VISIBLE);
-        }
         mChosenAnswer.setVisibility(View.VISIBLE);
         mFlipCardLayout.setVisibility(View.VISIBLE);
 
@@ -1765,8 +1731,6 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     protected SharedPreferences restorePreferences() {
         SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(getBaseContext());
 
-        mPrefHideDueCount = preferences.getBoolean("hideDueCount", false);
-        mPrefShowETA = preferences.getBoolean("showETA", true);
         mUseInputTag = preferences.getBoolean("useInputTag", false);
         // On newer Androids, ignore this setting, which should be hidden in the prefs anyway.
         mDisableClipboard = "0".equals(preferences.getString("dictionary", "0"));
@@ -1807,12 +1771,11 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     }
 
 
-    private void restoreCollectionPreferences() {
+    protected void restoreCollectionPreferences() {
 
         // These are preferences we pull out of the collection instead of SharedPreferences
         try {
             mShowNextReviewTime = getCol().getConf().getBoolean("estTimes");
-            mShowRemainingCardCount = getCol().getConf().getBoolean("dueCounts");
 
             // Dynamic don't have review options; attempt to get deck-specific auto-advance options
             // but be prepared to go with all default if it's a dynamic deck
@@ -1876,46 +1839,15 @@ public abstract class AbstractFlashcardViewer extends NavigationDrawerActivity i
     protected void updateScreenCounts() {
         if (mCurrentCard == null) return;
         ActionBar actionBar = getSupportActionBar();
-        int[] counts = mSched.counts(mCurrentCard);
 
         if (actionBar != null) {
             String title = Decks.basename(getCol().getDecks().get(mCurrentCard.getDid()).getString("name"));
             actionBar.setTitle(title);
-            if (mPrefShowETA) {
-                eta = mSched.eta(counts, false);
-                actionBar.setSubtitle(Utils.remainingTime(AnkiDroidApp.getInstance(), eta * 60));
-            }
         }
 
         if (!mPrefShowTopbar) {
             mTopBarLayout.setVisibility(View.GONE);
         }
-
-        newCount = new SpannableString(String.valueOf(counts[0]));
-        lrnCount = new SpannableString(String.valueOf(counts[1]));
-        revCount = new SpannableString(String.valueOf(counts[2]));
-        if (mPrefHideDueCount) {
-            revCount = new SpannableString("???");
-        }
-
-        switch (mSched.countIdx(mCurrentCard)) {
-            case Consts.CARD_TYPE_NEW:
-                newCount.setSpan(new UnderlineSpan(), 0, newCount.length(), 0);
-                break;
-            case Consts.CARD_TYPE_LRN:
-                lrnCount.setSpan(new UnderlineSpan(), 0, lrnCount.length(), 0);
-                break;
-            case Consts.CARD_TYPE_REV:
-                revCount.setSpan(new UnderlineSpan(), 0, revCount.length(), 0);
-                break;
-            default:
-                Timber.w("Unknown card type %s", mSched.countIdx(mCurrentCard));
-                break;
-        }
-
-        mTextBarNew.setText(newCount);
-        mTextBarLearn.setText(lrnCount);
-        mTextBarReview.setText(revCount);
     }
 
     /*
@@ -3745,6 +3677,9 @@ see card.js for available functions
  */
     // list of api that can be accessed
     private final String[] mApiList = {"toggleFlag", "markCard"};
+    public JavaScriptFunction javaScriptFunction() {
+        return new JavaScriptFunction();
+    }
 
     public class JavaScriptFunction {
 
@@ -3781,24 +3716,31 @@ see card.js for available functions
             return String.valueOf(apiStatusJson);
         }
 
+        // This method and the one belows return "default" values when there is no count nor ETA.
+        // Javascript may expect ETA and Counts to be set, this ensure it does not bug too much by providing a value of correct type
+        // but with a clearly incorrect value.
+        // It's overridden in the Reviewer, where those values are actually defined.
         @JavascriptInterface
         public String ankiGetNewCardCount() {
-            return newCount.toString();
+            return "-1";
         }
+
 
         @JavascriptInterface
         public String ankiGetLrnCardCount() {
-            return lrnCount.toString();
+            return "-1";
         }
+
 
         @JavascriptInterface
         public String ankiGetRevCardCount() {
-            return revCount.toString();
+            return "-1";
         }
+
 
         @JavascriptInterface
         public int ankiGetETA() {
-            return eta;
+            return -1;
         }
 
         @JavascriptInterface

--- a/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java
@@ -28,18 +28,23 @@ import android.os.Build;
 import android.os.Bundle;
 import android.os.Handler;
 import android.os.Message;
+import android.text.SpannableString;
+import android.text.style.UnderlineSpan;
 import android.view.KeyEvent;
 import android.view.Menu;
 import android.view.MenuItem;
 import android.view.SubMenu;
 import android.view.View;
 import android.view.ViewGroup;
+import android.webkit.JavascriptInterface;
 import android.widget.FrameLayout;
 import android.widget.LinearLayout;
+import android.widget.TextView;
 
 import androidx.annotation.DrawableRes;
 import androidx.annotation.NonNull;
 import androidx.annotation.VisibleForTesting;
+import androidx.appcompat.app.ActionBar;
 import androidx.core.app.ActivityCompat;
 import androidx.core.content.ContextCompat;
 import androidx.core.view.ActionProvider;
@@ -61,6 +66,7 @@ import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.Collection.DismissType;
 import com.ichi2.libanki.Consts;
 import com.ichi2.libanki.Decks;
+import com.ichi2.libanki.Utils;
 import com.ichi2.themes.Themes;
 import com.ichi2.utils.FunctionalInterfaces.Consumer;
 import com.ichi2.utils.Permissions;
@@ -85,6 +91,26 @@ public class Reviewer extends AbstractFlashcardViewer {
     private static final int ADD_NOTE = 12;
     private static final int REQUEST_AUDIO_PERMISSION = 0;
     private LinearLayout colorPalette;
+
+    // TODO: Consider extracting to ViewModel
+    // Card counts
+    private SpannableString newCount;
+    private SpannableString lrnCount;
+    private SpannableString revCount;
+
+    private TextView mTextBarNew;
+    private TextView mTextBarLearn;
+    private TextView mTextBarReview;
+
+    private boolean mPrefHideDueCount;
+
+    // ETA
+    private int eta;
+    private boolean mPrefShowETA;
+
+
+    // Preferences from the collection
+    private boolean mShowRemainingCardCount;
 
     private ActionButtons mActionButtons = new ActionButtons(this);
 
@@ -751,11 +777,54 @@ public class Reviewer extends AbstractFlashcardViewer {
     @Override
     protected SharedPreferences restorePreferences() {
         SharedPreferences preferences = super.restorePreferences();
+        mPrefHideDueCount = preferences.getBoolean("hideDueCount", false);
+        mPrefShowETA = preferences.getBoolean("showETA", true);
         this.mProcessor.setup();
         mBlackWhiteboard = preferences.getBoolean("blackWhiteboard", true);
         mPrefFullscreenReview = Integer.parseInt(preferences.getString("fullscreenMode", "0")) > 0;
         mActionButtons.setup(preferences);
         return preferences;
+    }
+
+    @Override
+    protected void updateScreenCounts() {
+        if (mCurrentCard == null) return;
+        ActionBar actionBar = getSupportActionBar();
+        int[] counts = mSched.counts(mCurrentCard);
+
+        if (actionBar != null) {
+            if (mPrefShowETA) {
+                eta = mSched.eta(counts, false);
+                actionBar.setSubtitle(Utils.remainingTime(AnkiDroidApp.getInstance(), eta * 60));
+            }
+        }
+
+
+        newCount = new SpannableString(String.valueOf(counts[0]));
+        lrnCount = new SpannableString(String.valueOf(counts[1]));
+        revCount = new SpannableString(String.valueOf(counts[2]));
+        if (mPrefHideDueCount) {
+            revCount = new SpannableString("???");
+        }
+
+        switch (mSched.countIdx(mCurrentCard)) {
+            case Consts.CARD_TYPE_NEW:
+                newCount.setSpan(new UnderlineSpan(), 0, newCount.length(), 0);
+                break;
+            case Consts.CARD_TYPE_LRN:
+                lrnCount.setSpan(new UnderlineSpan(), 0, lrnCount.length(), 0);
+                break;
+            case Consts.CARD_TYPE_REV:
+                revCount.setSpan(new UnderlineSpan(), 0, revCount.length(), 0);
+                break;
+            default:
+                Timber.w("Unknown card type %s", mSched.countIdx(mCurrentCard));
+                break;
+        }
+
+        mTextBarNew.setText(newCount);
+        mTextBarLearn.setText(lrnCount);
+        mTextBarReview.setText(revCount);
     }
 
     @Override
@@ -775,6 +844,31 @@ public class Reviewer extends AbstractFlashcardViewer {
     }
 
     @Override
+    protected void initLayout() {
+        mTextBarNew = (TextView) findViewById(R.id.new_number);
+        mTextBarLearn = (TextView) findViewById(R.id.learn_number);
+        mTextBarReview = (TextView) findViewById(R.id.review_number);
+
+        super.initLayout();
+
+        if (!mShowRemainingCardCount) {
+            mTextBarNew.setVisibility(View.GONE);
+            mTextBarLearn.setVisibility(View.GONE);
+            mTextBarReview.setVisibility(View.GONE);
+        }
+    }
+
+    @Override
+    protected void switchTopBarVisibility(int visible) {
+        super.switchTopBarVisibility(visible);
+        if (mShowRemainingCardCount) {
+            mTextBarNew.setVisibility(visible);
+            mTextBarLearn.setVisibility(visible);
+            mTextBarReview.setVisibility(visible);
+        }
+    }
+
+    @Override
     protected void onStop() {
         super.onStop();
 
@@ -791,6 +885,16 @@ public class Reviewer extends AbstractFlashcardViewer {
         if (mPrefWhiteboard) {
             setWhiteboardVisibility(mShowWhiteboard);
         }
+        if (mShowRemainingCardCount) {
+            mTextBarNew.setVisibility(View.VISIBLE);
+            mTextBarLearn.setVisibility(View.VISIBLE);
+            mTextBarReview.setVisibility(View.VISIBLE);
+        }
+    }
+
+    protected void restoreCollectionPreferences() {
+        super.restoreCollectionPreferences();
+        mShowRemainingCardCount = getCol().getConf().getBoolean("dueCounts");
     }
 
     @Override
@@ -1056,6 +1160,36 @@ public class Reviewer extends AbstractFlashcardViewer {
                 default:
                     return false;
             }
+        }
+    }
+
+    public ReviewerJavaScriptFunction javaScriptFunction() {
+        return new ReviewerJavaScriptFunction();
+    }
+
+    public class ReviewerJavaScriptFunction extends JavaScriptFunction {
+        @JavascriptInterface
+        @Override
+        public String ankiGetNewCardCount() {
+            return newCount.toString();
+        }
+
+        @JavascriptInterface
+        @Override
+        public String ankiGetLrnCardCount() {
+            return lrnCount.toString();
+        }
+
+        @JavascriptInterface
+        @Override
+        public String ankiGetRevCardCount() {
+            return revCount.toString();
+        }
+
+        @JavascriptInterface
+        @Override
+        public int ankiGetETA() {
+            return eta;
         }
     }
 }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java
@@ -750,10 +750,8 @@ public class Reviewer extends AbstractFlashcardViewer {
 
     @Override
     protected SharedPreferences restorePreferences() {
-        super.restorePreferences();
+        SharedPreferences preferences = super.restorePreferences();
         this.mProcessor.setup();
-        //Is this line necessary? Can we not use the return value from the call to super?
-        SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(getBaseContext());
         mBlackWhiteboard = preferences.getBoolean("blackWhiteboard", true);
         mPrefFullscreenReview = Integer.parseInt(preferences.getString("fullscreenMode", "0")) > 0;
         mActionButtons.setup(preferences);

--- a/AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.java
@@ -114,7 +114,7 @@ public class ReviewerTest extends RobolectricTest {
         addNoteUsingBasicModel("Hello", "World2");
 
         Reviewer reviewer = startReviewer();
-        JavaScriptFunction javaScriptFunction = reviewer.new JavaScriptFunction();
+        JavaScriptFunction javaScriptFunction = reviewer.javaScriptFunction();
 
 
         // The answer needs to be displayed to be able to get the time.
@@ -232,7 +232,7 @@ public class ReviewerTest extends RobolectricTest {
     private void assertCounts(Reviewer r, int newCount, int stepCount, int revCount) {
 
         List<String> countList = new ArrayList<>();
-        JavaScriptFunction jsApi = r.new JavaScriptFunction();
+        JavaScriptFunction jsApi = r.javaScriptFunction();
         countList.add(jsApi.ankiGetNewCardCount());
         countList.add(jsApi.ankiGetLrnCardCount());
         countList.add(jsApi.ankiGetRevCardCount());

